docs/contracts/using-for.rst

```diff
@@ -16,17 +16,17 @@ at contract level.
 
 The first part, ``A``, can be one of:
 
-- a list of file-level or library functions (``using {f, g, h, L.t} for uint;``) -
-  only those functions will be attached to the type.
+- a list of file-level or internal library functions (``using {f, g, h, L.t} for uint;``) -
```

Comment:
Wait, can't we also bind public library functions?

Also there is an error message that uses "attach" in DeclarationTypeChecker.cpp:488.

Maybe we should stick to "attach" for the user-facing parts?

Reply:
> Also there is an error message that uses "attach" in DeclarationTypeChecker.cpp:488.

https://github.com/ethereum/solidity/pull/13741#issuecomment-1327210307

Reply:
Also, I think binding is a more 'canon' (and therefore better) term for this type of feature instead of attach/ing.

Reply:
> Wait, can't we also bind public library functions?

True, we can. I didn't specify that because in many places docs are not precise about it and say internal when it really means the internal part of a function (i.e. both public and private qualify too). But I see that below we actually say "public" explicitly so I clarified it now.

But I also found two weird quirks, we maybe should address?
- `using L for uint` skips private functions while `using {L.f} for uint` does bind `f` even if it's private. **If you do that at file level, you can call a library function from outside of the library**
    ```solidity
    library L {
        function privateFunction(uint) private {}
    }

    using {L.privateFunction} for uint;

    contract C {
        function test() public {
            uint x;
            x.privateFunction();
        }
    }
    ```
- `using {L.f} for uint` binds the external part of the function. I.e. `x.f()` is an external call. If it works like this then what's the point of disallowing external library functions here?

Both of these look like bugs to me.

Reply:
> Also there is an error message that uses "attach" in DeclarationTypeChecker.cpp:488.

> Is [this](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/SyntaxChecker.cpp#L418) or [this](https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/DeclarationTypeChecker.cpp#L488) relevant?

Yeah. Added both of them.

Reply:
> Also, I think binding is a more 'canon' (and therefore better) term for this type of feature instead of attach/ing.

Yeah, I agree. Binding sounds more precise to me in a way. But I'd be fine with either, I just want a single term to refer to this consistently.

Reply:
Reported the two inconsistencies as bugs: #13764, #13765.

Though, since these are bugs, now I'm not sure if I should update the docs as if they weren't there (they behavior is not intentional) or keep it accurate and only change it when we fix them. I'm leaning towards leaving it accurate.

Reply:
If we document the wrong behaviour and then change it, it's no longer a bug, but a breaking change. I'd document the intended behaviour.